eng/scripts/test-matrix/test-matrix-functions.ps1

```diff
@@ -0,0 +1,317 @@
+Set-StrictMode -Version "4.0"
+
+function CreateDisplayName([string]$parameter, [hashtable]$displayNames) {
+    $name = $parameter
+
+    if ($displayNames[$parameter]) {
+        $name = $displayNames[$parameter]
+    }
+
+    return $name -replace "[\-\._]", ""
+}
+
+function GenerateMatrix([HashTable]$config, [string]$selectFromMatrixType) {
+    if ($selectFromMatrixType -eq "sparse") {
+        [Array]$dimensions = GetMatrixDimensions $config.matrix
+        $size = $dimensions[0]
+        [Array]$matrix = GenerateSparseMatrix $config.matrix $config.displayNames $size
+    } elseif ($selectFromMatrixType -eq "all") {
+        [Array]$matrix, $_ = GenerateFullMatrix $config.matrix $config.displayNames
+    } else {
+        throw "Matrix generator not implemented for selectFromMatrix: $($platform.selectFromMatrix)"
+    }
+
+    if ($config["exclude"]) {
+        [Array]$matrix = ProcessExcludes $matrix $config.exclude
+    }
+    if ($config["include"]) {
+        [Array]$matrix = ProcessIncludes $matrix $config.include $config.displayNames
+    }
+
+    return $matrix | Sort-Object -Property name
+}
+
+function ProcessExcludes([Array]$matrix, [Array]$excludes) {
+    $deleteKey = "%DELETE%"
+    $exclusionMatrix = @()
+
+    foreach ($exclusion in $excludes) {
+        $converted = ConvertToMatrixArrayFormat $exclusion
+        $full, $_ = GenerateFullMatrix $converted
+        $exclusionMatrix += $full
+    }
+
+    foreach ($element in $matrix) {
+        foreach ($exclusion in $exclusionMatrix) {
+            $match = MatrixElementMatch $element.parameters $exclusion.parameters
+            if ($match) {
+                $element.parameters[$deleteKey] = $true
+            }
+        }
+    }
+
+    return $matrix | Where-Object { !$_.parameters.Contains($deleteKey) }
+}
+
+function ProcessIncludes([Array]$matrix, [Array]$includes, [HashTable]$displayNames) {
+    foreach ($inclusion in $includes) {
+        $converted = ConvertToMatrixArrayFormat $inclusion
+        $full, $_ = GenerateFullMatrix $converted $displayNames
+        $matrix += $full
```

Comment:
Should we worry about including duplicate matrix entries?

Reply:
I played around with using a hashset to remove duplicates, but there were some limitations around it (I think preserving order), such that I decided to not solve the problem immediately. I can take another look at preventing duplicates in an efficient way.